File: tests/fixtures/scala/sample.scala
/**
 * Sample Scala file for extractor tests.
 */
package com.acme.users

import java.util.Optional
import com.acme.util.Strings
import com.acme.util._
import com.acme.util.{Formatter => Fmt, Parser}
import scala.collection.mutable

/** Base repository. */
abstract class BaseRepo[T](protected val session: Session) extends Repository[T] with Closeable {
  private val cache: Cache = new Cache()
  val MAX_RETRIES = 3

  /** Find one by id. */
  def find(id: String): Option[T]

  override def close(): Unit = session.close()
}

/** Finds users. */
class UserRepo(session: Session, strings: Strings) extends BaseRepo[User](session) {
  var count: Int = 0

  override def find(id: String): Option[User] = {
    val user = session.get(id)
    val v = new Validator(user)
    v.check()
    cache.put(user)
    val fmt: Fmt = Fmt()
    val key = System.getenv("API_TOKEN")
    Option(normalize(user))
  }

  private def normalize(u: User): User = u
}

trait Repository[T] extends AutoCloseable {
  def find(id: String): Option[T]
}

case class User(id: String, name: String)

object Registry extends Repository[User] {
  def find(id: String): Option[User] = None
  def lookup(id: String): User = users(id)
  val users = mutable.Map.empty[String, User]
}

type Handler = String => Unit

class UserRepoSpec extends AnyFunSuite {
  test("finds user") {
    val repo = new UserRepo(null, null)
    assert(repo.find("1").isEmpty)
  }
}

class UserRepoWordSpec extends AnyWordSpec {
  "UserRepo" should {
    "find users" in {
      val repo = new UserRepo(null, null)
      repo.find("1")
    }
  }
  it("also works") {
    Registry.lookup("x")
  }
}

File: tests/fixtures/kotlin/sample.kt
/**
 * Sample Kotlin file for extractor tests.
 */
package com.acme.users

import java.util.Optional
import com.acme.util.Strings
import com.acme.util.*
import com.acme.util.Formatter as Fmt
import org.springframework.web.bind.annotation.GetMapping

const val MAX_RETRIES = 3
val logger = Logger()
var counter: Int = 0

/** Base repository. */
abstract class BaseRepo<T>(protected val session: Session) : Repository<T>, Closeable {
    private val cache: Cache = Cache()

    /**
     * Find one by id.
     */
    abstract fun find(id: String): Optional<T>

    override fun close() {
        session.close()
    }

    companion object {
        const val NAME = "base"
        fun create(): BaseRepo<Any> = TODO()
    }
}

/**
 * Finds users.
 */
@Service
class UserRepo(session: Session, private val strings: Strings) : BaseRepo<User>(session) {
    val size: Int
        get() = 0

    override fun find(id: String): Optional<User> {
        val user = session.get(id)
        val v = Validator(user)
        v.check()
        cache.put(user)
        val fmt: Fmt = Fmt()
        val key = System.getenv("API_TOKEN")
        return Optional.ofNullable(normalize(user))
    }

    private fun normalize(u: User): User = u
}

interface Repository<T> {
    fun find(id: String): Optional<T>
}

data class User(val id: String, var name: String)

enum class Color(val code: String) {
    RED("r"), GREEN("g");
    fun label() = code
}

sealed class Result {
    object Empty : Result()
    data class Ok(val value: Int) : Result()
}

object Registry {
    fun lookup(id: String): User? = null
}

typealias Handler = (String) -> Unit

fun String.slug(): String = lowercase()

@RestController
class UserController(private val repo: UserRepo) {
    @GetMapping("/users/{id}")
    fun getUser(@PathVariable id: String): User = repo.find(id).get()

    @PostMapping(value = ["/users"])
    fun createUser(@RequestBody user: User) {
        repo.save(user)
    }
}

fun Application.module() {
    routing {
        get("/health") {
            call.respondText("ok")
        }
        post("/items") { }
    }
}

class UserRepoTest {
    @Test
    fun `finds user`() {
        val repo = UserRepo(Session(), Strings())
        assertNotNull(repo.find("1"))
    }
}
